libkbfs/file_data.go

```diff
@@ -290,16 +291,29 @@ func (fd *fileData) getLeafBlocksForOffsetRange(ctx context.Context,
 		respCh := make(chan resp, 1)
 		respChans = append(respChans, respCh)
 		eg.Go(func() error {
-			block, _, err := fd.getter(
-				groupCtx, fd.kmd, ptr, fd.file, blockReadParallel)
-			if err != nil {
-				return err
-			}
-			// Recurse down to the level of the child.
-			pfr, blocks, nextBlockOffset, err := fd.getLeafBlocksForOffsetRange(
-				groupCtx, ptr, block, startOff, endOff, prefixOk)
-			if err != nil {
-				return err
+			var pfr [][]parentBlockAndChildIndex
+			var blocks map[BlockPointer]*FileBlock
+			var nextBlockOffset int64
+			// We only need to fetch direct blocks if we've been asked
+			// to do so.
+			if getDirect || !ptr.IsDirect {
+				block, _, err := fd.getter(
+					groupCtx, fd.kmd, ptr, fd.file, blockReadParallel)
+				if err != nil {
+					return err
+				}
+
+				// Recurse down to the level of the child.
+				pfr, blocks, nextBlockOffset, err = fd.getBlocksForOffsetRange(
+					groupCtx, ptr, block, startOff, endOff, prefixOk, getDirect)
+				if err != nil {
+					return err
+				}
+			} else {
+				// We don't care about direct blocks, so leave the
+				// `blocks` map `nil`.
+				pfr = [][]parentBlockAndChildIndex{nil}
```

Comment:
This is the same thing `getBlocksForOffsetRange` does if the block is direct -- it's indicating that there is a single path to this block, but no parents have been appended to it yet.